### PR TITLE
#162162966: Users should be able to view and edit their profiles

### PR DESCRIPTION
**Pull request by billkabanga**
**What does this PR do?**

Creates UI functionality for users to view and edit the profiles.

**Description of the tasks to be completed?**
- Make the profile button functional. It should redirect user to profile page.
- Create profile page with edit functionality.

**How should this be manually tested?**
- Clone the repository
- git checkout ft-profile-view-edit-162162966
- Install dependencies using `yarn install`
- Run the app using `yarn start`
- The application should be running.
- Open browser and go to http://localhost:3000
- When logged in, click Profile and view/edit profile.

**What are the related pivotal tracker stories?**
#162162966

**Screenshots**

screen shot 2019-01-21 at 18 43 05
screen shot 2019-01-21 at 18 43 20
screen shot 2019-01-17 at 23 31 57
screen shot 2019-01-17 at 23 32 09

### CODE REVIEW

**Review by katunold — APPROVED** [review]
Well done bill

**Review by kizzanaome — APPROVED** [review]

**Comment by kizzanaome** [comment]
Please rebase and fix merge conflicts, otherwise  LGTM

**Review by arthurarty — APPROVED** [review]
lgtm

**Review by arthurarty — APPROVED** [review]
lgtm

File: src/components/profile/ProfileComponent.jsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React, { Component } from 'react';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import './profile.scss';
+
+
+export class ProfilePage extends Component {
+  render() {
+    const profile = this.props.profile;
```
Comment (matthewacha) [inline src/components/profile/ProfileComponent.jsx]: destructure this

File: src/components/profile/ProfileComponent.jsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    return (
+      <div className="container profile">
+        <div id="image">{profile.image === null ? <FontAwesomeIcon icon="user-circle" size={"10x"} color="#3F51B5"/>: <img className="rounded-circle" src={profile.image} alt="author profile..."/> }</div>
+        <h1>{profile.username}</h1>
+        <h6>{profile.bio}</h6>
+        <button type="button" className="btn btn-outline-primary" onClick={this.props.onClick}>Edit Profile</button>
```
Comment (matthewacha) [inline src/components/profile/ProfileComponent.jsx]: destructure `this.props.onClick`

File: src/components/profile/__tests__/components/profileEdit.test.jsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      bio: 'I love sports and tech',
+      image: null,
+      following: false,
+    },
+  };
+  const mockOnChange = jest.fn();
+  const mockOnClick = jest.fn();
+  const mockOnSubmit = jest.fn();
+  beforeEach(() => {
+    wrapper = mount(<ProfileEditComponent {...props} onChange={mockOnChange} onClick={mockOnClick} onSubmit={mockOnSubmit}/>);
```
Comment (matthewacha) [inline src/components/profile/__tests__/components/profileEdit.test.jsx:20]: pass these in the props as well.

File: src/components/profile/profileEditComponent.jsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import './profile.scss';
+
+export default class ProfileEdit extends Component {
+  render() {
+    const img = this.props.state
+    return (
+      <div className="container edit-form profile">
+        <form>
+          <div className="form-group profile-field">
+            <div>{img ? <img className="rounded-circle" src={img} alt="author profile..."/> : <FontAwesomeIcon icon="user-circle" size={"10x"} color="#3F51B5"/>}</div>
```
Comment (matthewacha) [inline src/components/profile/profileEditComponent.jsx]: this line is too long, fix it

File: src/components/profile/profileEditComponent.jsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import React, { Component } from 'react';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import './profile.scss';
```
Comment (matthewacha) [inline src/components/profile/profileEditComponent.jsx:4]: validate the props of these component

File: src/components/profile/profileEditComponent.jsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            </label>
+          </div>
+          <div className="form-group profile-field">
+            <input type="text" className="form-control mb-2 username" id="inlineFormInput" name="username" placeholder="Change your username here..." onChange={this.props.onChange}/>
+          </div>
+          <div className="form-group profile-field">
+            <textarea className="form-control bio" rows="5" id="comment" name="bio" placeholder="Add something to your bio here.." onChange={this.props.onChange}></textarea>
+          </div>
+          <button type="submit" className="btn btn-outline-primary save" onClick={this.props.onSubmit}>Save Changes</button>
+          <button type="submit" className="btn btn-outline-danger" onClick={this.props.onClick}>Cancel</button>
```
Comment (matthewacha) [inline src/components/profile/profileEditComponent.jsx]: destructure this.props.onClick

File: src/views/profile/profileEditView.jsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            image: result[0].secure_url,
+          },
+        });
+      },
+    );
+  }
+
+  handleSubmit = (event) => {
+    event.preventDefault();
+    this.props.postDataThunk('user/', this.state, editProfile, 'put');
```
Comment (matthewacha) [inline src/views/profile/profileEditView.jsx]: destructure this

File: src/views/profile/profileEditView.jsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    this.props.history.push('/profile');
+  }
+
+  render() {
+    if (this.props.profile) {
+      const { profile } = this.props.profile;
+      return (
+        <div>
+          <NavBar />
+          <ProfileEdit profile={profile} state={this.state.user.image} onChange={this.handleChange} onSubmit={this.handleSubmit} onClick={this.handleCancelClick} onClickChange={this.uploadImage}/>
```
Comment (matthewacha) [inline src/views/profile/profileEditView.jsx]: line is too long

File: src/views/profile/profileView.jsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import { connect } from 'react-redux';
+import ProfilePage from '../../components/profile/ProfileComponent';
+import { viewProfile } from '../../redux/actions/profileActions';
+import { getPrivateDataThunk } from '../../redux/thunks/index';
+import NavBar from '../../components/dashboard/NavBarComponent';
+
+export class ProfileView extends Component {
+
+  componentDidMount() {
+    this.props.getPrivateDataThunk('profiles/'.concat(localStorage.getItem('username')), viewProfile);
```
Comment (matthewacha) [inline src/views/profile/profileView.jsx]: destructure this

File: src/views/profile/profileView.jsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  handleClick = (event) => {
+    event.preventDefault();
+    this.props.history.push('/edit');
```
Comment (matthewacha) [inline src/views/profile/profileView.jsx]: destructure this

File: src/components/profile/__tests__/components/profileEdit.test.jsx (hunk 3 above)
Comment (billkabanga) [reply src/components/profile/__tests__/components/profileEdit.test.jsx:20]: I passed them.

**Review by matthewacha — APPROVED** [review]

**Review by matthewacha — APPROVED** [review]
good

**Pull request merged**